ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala

```diff
@@ -0,0 +1,416 @@
+        )
+      case PublicPackageUpload(_, sourceDescription, _, submissionId) =>
+        Map(
+          "updateSubmissionId" -> submissionId.getOrElse(""),
+          "updateSourceDescription" -> sourceDescription.getOrElse(""),
+        )
+      case PublicPackageUploadRejected(submissionId, _, rejectionReason) =>
+        Map(
+          "updateSubmissionId" -> submissionId,
+          "updateRejectionReason" -> rejectionReason,
+        )
+      case TransactionAccepted(optSubmitterInfo, transactionMeta, _, transactionId, _, _, _) =>
+        Map(
+          "updateTransactionId" -> transactionId,
+          "updateLedgerTime" -> transactionMeta.ledgerEffectiveTime.toInstant.toString,
+          "updateWorkflowId" -> transactionMeta.workflowId.getOrElse(""),
+          "updateSubmissionTime" -> transactionMeta.submissionTime.toInstant.toString,
+        ) ++ optSubmitterInfo
+          .map(info =>
+            Map(
+              "updateSubmitter" -> loggingContextPartiesValue(info.actAs),
+              "updateApplicationId" -> info.applicationId,
+              "updateCommandId" -> info.commandId,
+              "updateDeduplicateUntil" -> info.deduplicateUntil.toString,
+            )
+          )
+          .getOrElse(Map.empty)
+      case CommandRejected(_, submitterInfo, reason) =>
+        Map(
+          "updateSubmitter" -> loggingContextPartiesValue(submitterInfo.actAs),
+          "updateApplicationId" -> submitterInfo.applicationId,
+          "updateCommandId" -> submitterInfo.commandId,
+          "updateDeduplicateUntil" -> submitterInfo.deduplicateUntil.toString,
+          "updateRejectionReason" -> reason.description,
+        )
+    }
+
+  private def loggingContextPartiesValue(parties: List[Party]): String =
+    parties.mkString("[", ", ", "]")
+}
+
+class PipelinedExecuteUpdate(
+    private[indexer] val ledgerDao: LedgerDao,
+    private[indexer] val metrics: Metrics,
+    private[indexer] val participantId: v1.ParticipantId,
+)(implicit val executionContext: ExecutionContext, val loggingContext: LoggingContext)
+    extends ExecuteUpdate {
+
+  private val insertTransactionState: PreparedUpdate => Future[PreparedUpdate] = {
+    case offsetUpdate @ PreparedTransactionInsert(_, _, preparedInsert) =>
+      Timed.future(
+        metrics.daml.index.db.storeTransactionState,
+        ledgerDao.storeTransactionState(preparedInsert).map(_ => offsetUpdate),
+      )
+    case offsetUpdate => Future.successful(offsetUpdate)
+  }
+
+  private val insertTransactionEvents: PreparedUpdate => Future[PreparedUpdate] = {
+    case offsetUpdate @ PreparedTransactionInsert(_, _, preparedInsert) =>
+      Timed.future(
+        metrics.daml.index.db.storeTransactionEvents,
+        ledgerDao
+          .storeTransactionEvents(preparedInsert)
+          .map(_ => offsetUpdate),
+      )
+    case offsetUpdate => Future.successful(offsetUpdate)
+  }
+
+  private[indexer] val flow: ExecuteUpdateFlow =
```

Comment:
It might be worth adding something like `com.daml.metrics.Timed.timedAndTrackedFuture`, but for Akka streams. 

Another use case is measuring the number of elements being currently processed, or measuring the number of items in a `.buffer()` (cc @nmarton-da).

This is not something for this PR though.

Reply:
Added back the timer by threading the timer context through the stages. @rautenrieth-da A reason for which I did not want to time the entire flow is its inherent branching: we would be timing both metadata and pipelined transaction updates on the `store_ledger_entry` timer.